### ISSUE TITLE
Upcasting

### ISSUE THREAD

**Issue by dbrattli** [microsoft/pyright#1261]
**Describe the bug**

Would it be possible to get upcasting using a function to type check its arguments? Or more generally that pyright checks that arguments with `Type` annotation of a generic type e.g `Type[T]` are compatible with other arguments where the generic type `T` is also used.

**To Reproduce**

```python
from typing import TypeVar, Type
T = TypeVar("T")

class Base:
    pass

class Derived(Base):
    pass

class Other:
    pass

def upcast(type: Type[T], expr: T) -> T:
    return expr

d = Derived()
x = upcast(Other, d)
```

Would it be possible to get pyright to type check the last line and give error? Currently it will not detect that the function `upcast` is given two incompatible arguments. Here is how it looks in vscode today:

```python
upcast(Base, d)  # should not show error
upcast(Other, d)  # should show error
```

Currently no error will show even if Other is not a super type of the type of d:

<img width="365" alt="Screenshot 2020-12-10 at 08 17 56" src="https://user-images.githubusercontent.com/849479/101734237-42931780-3ac0-11eb-8820-0c7b88ee8e02.png">

**VS Code extension or command-line**

vscode extension: 1.1.94

**Comment by erictraut** [comment]
Pyright is doing the right thing here. It's correctly "solving" for the type of TypeVar T, and there are no type violations, so it's not displaying any errors.

I might be able to suggest a solution if I better understood what you're trying to do here. A derived class is always compatible with its base class, so there's no need to explicitly upcast.

If you want the static type checker to test whether an expression is of a type that is derived from `Base`, just do the following:
```python
# Will not show an error
y: Base = d

# Will show an error
z: Other = d
```

Does that address your use case?

**Comment by dbrattli** [comment]
I was trying to implement the `upcast` and `downcast` from F#. Upcast is perhaps not needed as you say, but good for symmetry. After reading https://github.com/python/typing/issues/565 I thought it should be possible to implement this using type annotations, and sort of expected pyright to give an error for `upcast(Other, d)` when `d` is not derived from `Other`.

**Comment by erictraut** [comment]
Yeah, there's no need for an `upcast` call. Upcasting is implicit in Python, as it is in most languages.
